File: generate/reference/duality-of-compilation/ocaml/lookup_tree.ml
type tree =
  | Leaf of int
  | Node of tree * tree

let rec create i n =
  if i < n then
    let t = create (i + 1) n in
    Node (t, t)
  else
    Leaf n

let rec lookup t =
  match t with
  | Leaf v -> v
  | Node (l, _) -> lookup l

let main n =
  lookup (create 0 n)

let () =
  match Sys.argv with
  | [| _ |] -> print_endline (string_of_int (main 10))
  | [| _; s |] -> (
      match int_of_string_opt s with
      | Some n -> print_endline (string_of_int (main n))
      | None -> print_endline "argument is not a number")
  | _ -> print_endline "arguments are too many"

File: generate/reference/duality-of-compilation/rust/lookup_tree.rs
use std::env;
use std::rc::Rc;

#[derive(Clone)]
enum Tree {
    Leaf(isize),
    Node(Rc<Tree>, Rc<Tree>),
}

fn create(i: isize, n: isize) -> Tree {
    if i < n {
        let tree = Rc::new(create(i + 1, n));
        Tree::Node(tree.clone(), tree)
    } else {
        Tree::Leaf(n)
    }
}

fn lookup(tree: &Tree) -> isize {
    match tree {
        Tree::Leaf(value) => *value,
        Tree::Node(left, _) => lookup(left),
    }
}

fn run(n: isize) -> isize {
    lookup(&create(0, n))
}

fn main() {
    let args: Vec<String> = env::args().collect();

    let arg = match args.get(1) {
        None => 10,
        Some(val) => match val.parse::<isize>() {
            Ok(val) => val,
            Err(_) => {
                println!("argument is not a number");
                return;
            }
        },
    };

    println!("{}", run(arg));
}
